src/EscapeAnalysis.jl

```diff
@@ -252,30 +290,29 @@ function find_escapes(ir::IRCode, nargs::Int)
         for pc in nstmts:-1:1
             stmt = stmts.inst[pc]
 
-            # inliner already computed effect-freeness of this statement (whether it may throw or not)
-            # and if it may throw, we conservatively escape all the arguments
+            # we escape statements with the `ThrownEscape` property using the effect-freeness
+            # information computed by the inliner
             is_effect_free = stmts.flag[pc] & IR_FLAG_EFFECT_FREE ≠ 0
 
             # collect escape information
             if isa(stmt, Expr)
                 head = stmt.head
-                if head === :call # TODO implement more builtins, make them more accurate
+                if head === :call
+                    has_changes = escape_call!(stmt.args, pc, state, ir, changes)
                     if !is_effect_free
-                        # TODO we can have a look at builtins.c and limit the escaped arguments if any of them are not captured in the thrown exception
-                        add_changes!(stmt.args[2:end], ir, Escape(), changes)
+                        add_changes!(stmt.args, ir, ThrownEscape(), changes)
                     else
-                        escape_call!(stmt.args, pc, state, ir, changes) || continue
+                        has_changes || continue
                     end
                 elseif head === :invoke
                     linfo = first(stmt.args)
                     escapes_for_call = get(GLOBAL_ESCAPE_CACHE, linfo, nothing)
                     if isnothing(escapes_for_call)
-                        add_changes!(stmt.args[3:end], ir, Escape(), changes)
+                        add_changes!(stmt.args[3:end], ir, AllEscape(), changes)
                     else
                         for (arg, info) in zip(stmt.args[2:end], escapes_for_call.arguments)
-                            if info === ReturnEscape()
-                                info = NoEscape()
-                            end
+                            # reinterpret the escape information in the context of this analysis frame
+                            info = EscapeLattice(true, false, info.ThrownEscape, info.GlobalEscape)
```

Comment:
this seems like it should have a helper function for this operation ("from_interproceedural(info, state.ssavalues[pc])")

Reply:
Sounds reasonable, but how `state.ssavalues[pc]` is supposed to be used for ?
Propagate `state.ssavalues[pc]` to each call argument, because the return value might be an alias to any of those call arguments, right ?

Reply:
@TH3CHARLie thanks for your idea on `isbitstype`, it actually seems to help us keep analysis accuracy even if we haven't implemented proper alias analysis, e.g.:
```julia
function f_no_return_escape(a)
    println("hi") # prevent inlining
    return nothing
end
let
    obj = Ref("foo")              # still be `NoEscape`
    ret = f_no_return_escape(obj)
    return ret                    # we don't impose `ReturnEscape` on `ret` because it's just `Const(nothing)`
end
```

Reply:
`ReturnEscape()` as the first argument there would mean that the escape information for that argument is exactly `state.ssavalues[pc]`. Others may also need to be translated in the future, but currently we don't (afaik) track other info.

Reply:
Ah, I got the point.

Yes, my confusion came from the fact that currently `EscapeLattice` doesn't distinguish "return to the caller as return value" vs. "return to the caller because it's visible from the caller", and for the latter case, we even don't need propagate `state.ssavalues[pc]`.

#37 will extend `EscapeLattice.ReturnEscape` so that it can recognize the difference, and I'd like to make `from_interprocedural` more accurate using those difference in that PR.